diff --git a/tools/python/transform/transformer.py b/tools/python/transform/transformer.py
index bb9154f8..3a0a2722 100644
--- a/tools/python/transform/transformer.py
+++ b/tools/python/transform/transformer.py
@@ -961,7 +961,7 @@ class Transformer(base_converter.ConverterInterface):
                 if consumer_op.type == MaceOp.Activation.name \
                         and ConverterUtil.get_arg(
                             consumer_op,
-                            MaceKeyword.mace_activation_type_str).s != 'PRELU':
+                            MaceKeyword.mace_activation_type_str).s != b'PRELU':
                     print("Fold activation: %s(%s)" % (op.name, op.type))
                     op.name = consumer_op.name
                     op.output[0] = consumer_op.output[0]
