diff --git a/tools/bazel_build_standalone_lib.sh b/tools/bazel_build_standalone_lib.sh
index 3fa74f3..0d44e42 100755
--- a/tools/bazel_build_standalone_lib.sh
+++ b/tools/bazel_build_standalone_lib.sh
@@ -115,7 +115,7 @@ $(echo "$1" | cut -d '=' -f -1)"
   esac
 done
 
-if [[ "${enable_apu}" == true || "${abi}" != armeabi-v7a || "${abi}" != arm64-v8a ]];then
+if [[ "${enable_apu}" == true || ("${abi}" != armeabi-v7a && "${abi}" != arm64-v8a) ]];then
   enable_rpcmem=false
 fi
 
